## **Extracting valuable business insights from the data of a fictional company, MovieNow**
*Data-Driven decision making in SQL*

---------------

The objective of this project is to provide decision-making support to the fictional movie rental company MovieNow using SQL. By extracting relevant information from the database, the project focuses on measuring the success of MovieNow. Key metrics for evaluation include revenue generation, average movie ratings as an indicator of customer satisfaction, and the number of active customers as a customer engagement metric.

Objectives of Data-driven decision making:

Data-driven decision making achieves short-term and long-term goals. We can extract valuable information from data to support operational short-term decisions. For example, the popularity of certain actors helps MovieNow decide whether to purchase certain movies. Also, last month's revenue can be important information supporting decisions regarding short-term investments. For long-term decisions, data-driven support can provide information regarding customer growth and successes in certain regions in the past, which can inform company decisions regarding when and where the market can be expanded in the future. Also, knowing the long-term development of overall revenue helps MovieNow plan long-term investments. *source: DataCamp*


KPIs: Key Performance Indicators

Key performance indicators help a company define and monitor success. Revenue is a trivial indicator of success. For MovieNow, this is calculated as the sum of the price for rented movies. For the subdivision customer relations management, the KPI 'customer satisfaction' could be quantified by the average rating of all movies or the KPI 'customer engagement' as the number of active customers in a certain time period. *source: DataCamp*



*To connect PostgreSql with Jupyter Notebook, follow these steps:*
1. Install Required Packages: ipython-sql and psycopg2 

2. Load the SQL Extension: %load_ext sql 

3. Connect to PostgreSQL: %sql postgresql://username:password@host:port/database_name *Replace your password*

In [ ]:
pip install psycopg2
pip install ipython-sql
%load_ext sql
%sql postgresql://username:password@host:port/database_name

In [ ]:
%sql postgresql://postgres:311279Ma@localhost/movie_now

The database belongs to a fictional movie rental company called MovieNow, which contains information about consumers, 
movie ratings, actor details, and more. The data were downloaded from the course 'Data-Driven Decision Making in SQL' on DataCamp. 

Tables in the Movie_now database: 

1. 'customers' 
	
2. 'movies' : contains all movies available on the online platform.
	
3. 'renting' 

4. 'actors'

5. 'actsin'

### **Business Case:**
The goal is to help management make informed and data driven decision 

---------------

*MovieNow considers to invest money in new movies, based on the data in wich kind of movies the new budget should be invested?*

Considering that it is more expensive for MovieNow to make recently produced movies available compared to older ones, it is a good starting point to investigate customer preferences based on the year of movie release. 

- Do customers give higher ratings to recently produced movies compared to older ones?

- Is there a difference in preferences across countries? 

To answer these questions, we will focus only on records of movies with at least 4 ratings and rentals since 2018-04-01. We will count the number of movie rentals, the number of different movies, and calculate the average rating.

In [85]:
%%sql 
SELECT c.country, 
    m.year_of_release,
    COUNT(*) AS n_rentals, 
    COUNT(DISTINCT r.movie_id) AS n_movies,
    ROUND(AVG(rating),2) AS avg_rating
FROM renting AS r
LEFT JOIN customers AS c 
ON c.customer_id= r.customer_id
LEFT JOIN movies AS m 
ON m.movie_id= r.movie_id
WHERE r.movie_id IN (
                    SELECT movie_id
                    FROM renting 
                    GROUP BY movie_id 
                    HAVING COUNT(rating)>= 4)
AND r.date_renting>='2018-04-01'
GROUP BY ROLLUP(m.year_of_release, c.country)
ORDER BY c.country, m.year_of_release; 

 * postgresql://postgres:***@localhost/movie_now
131 rows affected.


[('Austria', 2002, 1, 1, Decimal('7.00')),
 ('Austria', 2004, 1, 1, None),
 ('Austria', 2007, 1, 1, Decimal('6.00')),
 ('Austria', 2011, 1, 1, Decimal('6.00')),
 ('Belgium', 2001, 5, 3, Decimal('10.00')),
 ('Belgium', 2002, 1, 1, None),
 ('Belgium', 2003, 3, 2, Decimal('10.00')),
 ('Belgium', 2005, 1, 1, None),
 ('Belgium', 2006, 2, 2, Decimal('7.00')),
 ('Belgium', 2007, 3, 3, Decimal('9.33')),
 ('Belgium', 2009, 1, 1, Decimal('6.00')),
 ('Belgium', 2010, 1, 1, Decimal('8.00')),
 ('Belgium', 2011, 1, 1, None),
 ('Denmark', 2001, 2, 1, Decimal('8.00')),
 ('Denmark', 2002, 1, 1, Decimal('6.00')),
 ('Denmark', 2003, 2, 2, Decimal('7.50')),
 ('Denmark', 2004, 3, 2, Decimal('10.00')),
 ('Denmark', 2006, 2, 2, Decimal('8.00')),
 ('Denmark', 2007, 2, 2, Decimal('7.00')),
 ('Denmark', 2009, 1, 1, Decimal('10.00')),
 ('Denmark', 2011, 2, 1, Decimal('9.00')),
 ('Denmark', 2013, 1, 1, Decimal('9.00')),
 ('France', 2001, 4, 3, Decimal('6.00')),
 ('France', 2002, 6, 4, Decimal('8.33')),
 ('France', 2003, 5, 4, Decimal('9.00')),
 ('France', 2004, 5, 3, Decimal('8.50')),
 ('France', 2005, 6, 3, Decimal('6.25')),
 ('France', 2006, 2, 2, None),
 ('France', 2007, 5, 4, Decimal('7.50')),
 ('France', 2008, 2, 2, Decimal('8.50')),
 ('France', 2009, 2, 1, Decimal('8.00')),
 ('France', 2010, 7, 3, Decimal('8.00')),
 ('France', 2012, 3, 3, Decimal('9.00')),
 ('Great Britan', 2001, 2, 1, Decimal('10.00')),
 ('Great Britan', 2003, 3, 2, Decimal('4.50')),
 ('Great Britan', 2004, 6, 5, Decimal('7.50')),
 ('Great Britan', 2005, 1, 1, None),
 ('Great Britan', 2006, 2, 2, Decimal('6.00')),
 ('Great Britan', 2007, 2, 2, Decimal('10.00')),
 ('Great Britan', 2008, 1, 1, Decimal('9.00')),
 ('Great Britan', 2009, 1, 1, Decimal('8.00')),
 ('Great Britan', 2010, 2, 1, None),
 ('Great Britan', 2012, 2, 1, Decimal('8.00')),
 ('Great Britan', 2013, 2, 2, Decimal('7.50')),
 ('Hungary', 2001, 3, 2, Decimal('8.33')),
 ('Hungary', 2002, 2, 2, Decimal('7.00')),
 ('Hungary', 2003, 5, 4, Decimal('8.50')),
 ('Hungary', 2004, 5, 3, Decimal('7.25')),
 ('Hungary', 2005, 1, 1, Decimal('10.00')),
 ('Hungary', 2006, 4, 3, Decimal('6.00')),
 ('Hungary', 2007, 1, 1, Decimal('8.00')),
 ('Hungary', 2010, 3, 3, Decimal('6.33')),
 ('Hungary', 2012, 1, 1, Decimal('10.00')),
 ('Hungary', 2013, 1, 1, Decimal('5.00')),
 ('Hungary', 2014, 2, 1, Decimal('8.00')),
 ('Italy', 2001, 4, 3, Decimal('9.00')),
 ('Italy', 2002, 1, 1, Decimal('8.00')),
 ('Italy', 2003, 3, 2, Decimal('8.00')),
 ('Italy', 2004, 10, 4, Decimal('7.86')),
 ('Italy', 2005, 3, 2, Decimal('9.00')),
 ('Italy', 2006, 7, 3, Decimal('9.00')),
 ('Italy', 2007, 7, 4, Decimal('8.75')),
 ('Italy', 2008, 5, 3, Decimal('7.50')),
 ('Italy', 2010, 8, 5, Decimal('7.67')),
 ('Italy', 2011, 6, 2, Decimal('9.00')),
 ('Italy', 2012, 9, 5, Decimal('7.25')),
 ('Poland', 2001, 2, 2, Decimal('9.00')),
 ('Poland', 2002, 3, 2, Decimal('7.67')),
 ('Poland', 2003, 2, 1, Decimal('8.00')),
 ('Poland', 2004, 4, 3, Decimal('9.50')),
 ('Poland', 2005, 2, 2, Decimal('6.00')),
 ('Poland', 2006, 4, 3, Decimal('8.00')),
 ('Poland', 2007, 3, 3, Decimal('8.00')),
 ('Poland', 2008, 2, 1, Decimal('3.00')),
 ('Poland', 2010, 8, 4, Decimal('8.50')),
 ('Poland', 2011, 2, 1, Decimal('8.00')),
 ('Poland', 2012, 4, 3, Decimal('7.33')),
 ('Poland', 2013, 1, 1, None),
 ('Slovenia', 2001, 2, 1, Decimal('7.00')),
 ('Slovenia', 2002, 2, 2, None),
 ('Slovenia', 2003, 3, 3, Decimal('8.00')),
 ('Slovenia', 2004, 2, 2, Decimal('8.50')),
 ('Slovenia', 2005, 1, 1, Decimal('6.00')),
 ('Slovenia', 2006, 3, 3, Decimal('7.67')),
 ('Slovenia', 2007, 4, 4, Decimal('9.00')),
 ('Slovenia', 2008, 1, 1, None),
 ('Slovenia', 2009, 2, 1, Decimal('10.00')),
 ('Slovenia', 2010, 4, 3, Decimal('8.75')),
 ('Slovenia', 2012, 2, 2, Decimal('8.50')),
 ('Slovenia', 2013, 1, 1, None),
 ('Slovenia', 2014, 1, 1, Decimal('6.00')),
 ('Spain', 2001, 6, 4, Decimal('7.25')),
 ('Spain', 2002, 6, 3, Decimal('7.50')),
 ('Spain', 2003, 1, 1, None),
 ('Spain', 2004, 5, 4, Decimal('7.50')),
 ('Spain', 2005, 5, 2,

In the final lines of the resulting table, we observe the total aggregation, which includes 333 movie rentals, 50 different movies, and an average rating of 7.9. The subsequent rows represent aggregations for each year of release, arranged in ascending order. While a more detailed statistical analysis could be conducted to explore potential trends, it is evident from an initial observation that **there is no distinct pattern indicating better ratings for movies with a more recent year of release.**

To determine the type of movies in which the new budget should be invested we can consider allocating funds to movies of the highest-rated genres.

- Do customers have higher ratings for some movie genres rather than others?

In [86]:
%%sql
SELECT genre,
        AVG(rating) AS avg_rating,
        COUNT(rating) AS n_rating,
        COUNT(*) AS n_rentals,
        COUNT(DISTINCT m.movie_id) AS n_movies 
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 3 )
AND r.date_renting >= '2018-01-01'
GROUP BY genre
ORDER BY n_rating DESC;

 * postgresql://postgres:***@localhost/movie_now
8 rows affected.


[('Drama', Decimal('7.7482517482517483'), 143, 245, 34),
 ('Science Fiction & Fantasy', Decimal('8.2765957446808511'), 47, 70, 10),
 ('Comedy', Decimal('7.9500000000000000'), 20, 31, 5),
 ('Mystery & Suspense', Decimal('7.4285714285714286'), 7, 19, 3),
 ('Other', Decimal('8.4285714285714286'), 7, 16, 2),
 ('Action & Adventure', Decimal('8.7142857142857143'), 7, 9, 2),
 ('Animation', Decimal('7.8333333333333333'), 6, 10, 2),
 ('Art House & International', Decimal('8.5000000000000000'), 4, 5, 1)]

Based on the resulting table we found out that Action & Adventure has the highest rating.

Also, we can get some insights about if there is some customer preferences for certain actors based in the actors nationality and gender. 

- Do customers give higher ratings to actors of certain nationalities and genders compared to others?

In [88]:
%%sql
SELECT a.nationality,
        a.gender,
        ROUND(AVG(r.rating),2) AS avg_rating,
        COUNT(r.rating) AS n_rating,
        COUNT(*) AS n_rentals,
        COUNT(DISTINCT a.actor_id) AS n_actors
FROM renting AS r
LEFT JOIN actsin AS ai
ON ai.movie_id = r.movie_id
LEFT JOIN actors AS a
ON ai.actor_id = a.actor_id
WHERE r.movie_id IN (
                    SELECT movie_id
                    FROM renting
                    GROUP BY movie_id
                    HAVING COUNT(rating) >= 4)
AND r.date_renting >= '2018-04-01'
GROUP BY CUBE (nationality,gender);

 * postgresql://postgres:***@localhost/movie_now
36 rows affected.


[('Argentina', 'male', Decimal('8.50'), 4, 5, 1),
 ('Argentina', None, Decimal('8.50'), 4, 5, 1),
 ('Australia', 'female', Decimal('8.67'), 3, 5, 1),
 ('Australia', 'male', Decimal('7.45'), 11, 17, 3),
 ('Australia', None, Decimal('7.71'), 14, 22, 4),
 ('Austria', 'male', Decimal('8.50'), 2, 6, 1),
 ('Austria', None, Decimal('8.50'), 2, 6, 1),
 ('British', 'female', Decimal('7.83'), 54, 78, 3),
 ('British', 'male', Decimal('8.11'), 114, 175, 9),
 ('British', None, Decimal('8.02'), 168, 253, 12),
 ('Canada', 'female', Decimal('8.00'), 5, 5, 1),
 ('Canada', 'male', Decimal('8.43'), 7, 13, 2),
 ('Canada', None, Decimal('8.25'), 12, 18, 3),
 ('Ireland', 'male', Decimal('7.71'), 7, 16, 2),
 ('Ireland', None, Decimal('7.71'), 7, 16, 2),
 ('Israel', 'female', Decimal('8.25'), 4, 8, 1),
 ('Israel', None, Decimal('8.25'), 4, 8, 1),
 ('Italy', 'female', Decimal('9.00'), 4, 4, 1),
 ('Italy', None, Decimal('9.00'), 4, 4, 1),
 ('Northern Ireland', 'male', Decimal('9.00'), 4, 4, 1),
 ('Northern Ireland', None, Decimal('9.00'), 4, 4, 1),
 ('Puerto Rico', 'male', Decimal('7.20'), 5, 7, 1),
 ('Puerto Rico', None, Decimal('7.20'), 5, 7, 1),
 ('Somalia', 'male', Decimal('7.00'), 3, 5, 1),
 ('Somalia', None, Decimal('7.00'), 3, 5, 1),
 ('South Africa', 'female', Decimal('8.21'), 14, 21, 1),
 ('South Africa', None, Decimal('8.21'), 14, 21, 1),
 ('Spain', 'female', Decimal('8.50'), 4, 5, 1),
 ('Spain', 'male', Decimal('8.86'), 7, 12, 2),
 ('Spain', None, Decimal('8.73'), 11, 17, 3),
 ('USA', 'female', Decimal('7.55'), 121, 212, 27),
 ('USA', 'male', Decimal('7.91'), 242, 401, 42),
 ('USA', None, Decimal('7.79'), 363, 613, 69),
 (None, None, Decimal('7.90'), 615, 999, 101),
 (None, 'female', Decimal('7.75'), 209, 338, 36),
 (None, 'male', Decimal('7.98'), 406, 661, 65)]

Based on the resulting table from analyzing 999 records, it appears that there is a preference for British and USA movies, with 253 and 614 rentals respectively. Additionally, it was observed that male actors are slightly preferred, as in 661 out of the 999 analyzed movies, the actor was male.

*With these insights into customer preferences, the company can make informed, data-driven decisions.*

>

**In addition, with the data provided it is possible to extract some important information.**

---------------

- Get important KPIs (the revenue coming from movie rentals, the number of movie rentals and the number of active customers) for the performance of the company in 2018

In [90]:
%%sql 
SELECT SUM(m.renting_price) AS Revenue, 
        COUNT(*) AS movie_rentals, 
        COUNT(DISTINCT r.customer_id) AS active_customers
FROM renting AS r
LEFT JOIN movies AS m
ON r.movie_id = m.movie_id
-- Only look at movie rentals in 2018
WHERE date_renting BETWEEN '2018-01-01' AND '2018-12-31';

 * postgresql://postgres:***@localhost/movie_now
1 rows affected.


[(Decimal('658.02'), 298, 93)]

- Get the number of countries where MovieNow has customers 

In [23]:
%%sql 
SELECT COUNT(DISTINCT country)   
FROM customers;

 * postgresql://postgres:***@localhost/movie_now
1 rows affected.


[(11,)]

- Get summary statistics for the prerefences of the costumers in the movie 'Bridget Jones - The Edge of Reason'. (movie_ID=25) 

In [31]:
%%sql  
SELECT MIN(rating) min_rating, 
    MAX(rating) max_rating, 
    Round(AVG(rating),2) avg_rating,
    Count(rating) number_ratings 
FROM renting
WHERE movie_id = 25;

 * postgresql://postgres:***@localhost/movie_now
1 rows affected.


[(5, 10, Decimal('7.50'), 8)]

- Gey the average renting price by movie genre 

In [62]:
%%sql 
SELECT genre,
       ROUND(AVG(renting_price),2) AS avg_price,
       COUNT(*) AS number_movies
FROM movies 
GROUP BY genre
ORDER BY avg_price DESC;

 * postgresql://postgres:***@localhost/movie_now
8 rows affected.


[('Other', Decimal('2.59'), 2),
 ('Art House & International', Decimal('2.59'), 1),
 ('Action & Adventure', Decimal('2.46'), 3),
 ('Comedy', Decimal('2.29'), 10),
 ('Science Fiction & Fantasy', Decimal('2.24'), 11),
 ('Mystery & Suspense', Decimal('2.17'), 6),
 ('Drama', Decimal('2.15'), 36),
 ('Animation', Decimal('1.79'), 2)]

- Look at the average movie ratings for customers with more than 7 movies rentals 

In [89]:
%%sql 
SELECT customer_id,
      ROUND(AVG(rating),2) AS avg_renting,  
      COUNT(rating) AS no_ratings,  
      COUNT(date_renting) AS movie_rentals  
FROM renting
GROUP BY customer_id
HAVING COUNT(date_renting) > 7 
ORDER BY avg(rating)
LIMIT 3;

 * postgresql://postgres:***@localhost/movie_now
3 rows affected.


[(104, Decimal('6.25'), 4, 8),
 (28, Decimal('6.71'), 7, 11),
 (113, Decimal('7.00'), 7, 15)]

- Get the top 10 money spent per customer

In [91]:
%%sql 
SELECT rm.customer_id,
        SUM(rm.renting_price) AS money_spent_by_consumer 
FROM (SELECT r.customer_id,
             m.renting_price
      FROM renting AS r
     LEFT JOIN movies AS m
     ON r.movie_id=m.movie_id) AS rm 
GROUP BY rm.customer_id
ORDER BY SUM(rm.renting_price) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost/movie_now
10 rows affected.


[(21, Decimal('31.56')),
 (113, Decimal('31.45')),
 (49, Decimal('30.37')),
 (114, Decimal('24.29')),
 (108, Decimal('24.00')),
 (25, Decimal('24.00')),
 (92, Decimal('23.89')),
 (28, Decimal('21.89')),
 (65, Decimal('21.61')),
 (111, Decimal('21.30'))]

- Get the top 10 movies income

In [92]:
%%sql 
SELECT rm.title, 
       SUM(rm.renting_price) AS income_movie
FROM
       (SELECT m.title,  
               m.renting_price
       FROM renting AS r
       LEFT JOIN movies AS m
       ON r.movie_id=m.movie_id) AS rm
Group by (rm.title)
ORDER BY income_movie DESC
LIMIT 10;

 * postgresql://postgres:***@localhost/movie_now
10 rows affected.


[('Bridget Jones - The Edge of Reason', Decimal('37.57')),
 ('Fair Game', Decimal('34.68')),
 ('The Kingdom', Decimal('31.35')),
 ('Two for the Money', Decimal('30.69')),
 ('Simone', Decimal('29.59')),
 ('Django Unchained', Decimal('29.59')),
 ('What Women Want', Decimal('28.49')),
 ('Monster', Decimal('27.17')),
 ('The Two Towers', Decimal('26.29')),
 ('A Good Woman', Decimal('26.01'))]

- Which is the favorite movie for all customers born in the 70s?

In [94]:
%%sql 
SELECT m.title, 
COUNT(*) AS views,
ROUND(AVG(r.rating),2) AS avg_rating
FROM renting AS r
LEFT JOIN customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE c.date_of_birth BETWEEN '1970-01-01' AND '1979-12-31'
GROUP BY m.title
HAVING COUNT(*) <> 1 AND AVG(r.rating) IS NOT NULL
ORDER BY AVG(r.rating) DESC 
LIMIT 10;

 * postgresql://postgres:***@localhost/movie_now
10 rows affected.


[('The Fighter', 4, Decimal('10.00')),
 ("I'm Not There", 2, Decimal('10.00')),
 ("One Night at McCool's", 2, Decimal('10.00')),
 ('No Country for Old Men', 3, Decimal('10.00')),
 ('Django Unchained', 4, Decimal('10.00')),
 ('The Fellowship of the Ring', 2, Decimal('10.00')),
 ('Secondhand Lions', 4, Decimal('9.67')),
 ("Winter's Bone", 2, Decimal('9.50')),
 ("Fool's Gold", 4, Decimal('9.33')),
 ('Burn After Reading', 2, Decimal('9.00'))]

- Get the movies with rating above average

In [69]:
%%sql 
SELECT title 
FROM movies
WHERE movie_id in 
                (SELECT movie_id
                 FROM renting
                 GROUP BY movie_id
                 HAVING AVG(rating) > 
                                     (SELECT AVG(rating)
                                      FROM renting))
LIMIT 10;

 * postgresql://postgres:***@localhost/movie_now
10 rows affected.


[('What Women Want',),
 ('The Fellowship of the Ring',),
 ("Harry Potter and the Philosopher's Stone",),
 ('The Royal Tenenbaums',),
 ('Waking Up in Reno',),
 ("11'09''01 September 11",),
 ('Harry Potter and the Chamber of Secrets',),
 ('25th Hour',),
 ('The Human Stain',),
 ('Secondhand Lions',)]

- Get the name of customers who rented fewer than 5 movies

In [73]:
%%sql 
SELECT *
FROM customers as c
WHERE 5 > 
        (SELECT count(*)
         FROM renting as r
         WHERE r.customer_id = c.customer_id)
LIMIT 5;

 * postgresql://postgres:***@localhost/movie_now
5 rows affected.


[(2, 'Wolfgang Ackermann', 'Austria', 'male', datetime.date(1971, 11, 17), datetime.date(2018, 10, 15)),
 (3, 'Daniela Herzog', 'Austria', 'female', datetime.date(1974, 8, 7), datetime.date(2019, 2, 14)),
 (4, 'Julia Jung', 'Austria', 'female', datetime.date(1991, 1, 4), datetime.date(2017, 11, 22)),
 (5, 'Juliane Kirsch', 'Austria', 'female', datetime.date(1977, 3, 1), datetime.date(2018, 12, 16)),
 (6, 'Rowanne Couperus', 'Belgium', 'female', datetime.date(1994, 4, 5), datetime.date(2018, 8, 26))]

- Identify customers who were not satisfied with movies they watched on MovieNow. Customers who gave low ratings, report a list of customers with minimum rating smaller than 4

In [75]:
%%sql 
SELECT *
FROM customers as c
WHERE 4 >
        (SELECT MIN(rating)
         FROM renting AS r
         WHERE r.customer_id = c.customer_id)

 * postgresql://postgres:***@localhost/movie_now
4 rows affected.


[(28, 'Sidney Généreux', 'France', 'male', datetime.date(1980, 12, 1), datetime.date(2017, 2, 4)),
 (41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12)),
 (86, 'Albin Jaworski', 'Poland', 'male', datetime.date(1984, 5, 1), datetime.date(2017, 12, 15)),
 (120, 'Robin J. Himes', 'USA', 'male', datetime.date(1988, 11, 30), datetime.date(2018, 8, 6))]

---